core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,197 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to preload image into AssetsBackendApiService's cache.
+ */
+
+oppia.factory('ImagePreloaderService', [
+  '$q', '$uibModal', 'ExplorationContextService', 'AssetsBackendApiService',
+  'ExplorationPlayerStateService', 'UrlInterpolationService',
+  'ComputeGraphService', 'ExtractImageFilenamesFromStateService',
+  function($q, $uibModal, ExplorationContextService, AssetsBackendApiService,
+      ExplorationPlayerStateService, UrlInterpolationService,
+      ComputeGraphService, ExtractImageFilenamesFromStateService) {
+    var MAX_NUM_IMAGE_FILES_TO_DOWNLOAD_SIMULTANEOUSLY = 3;
+
+    var _filenamesOfImageCurrentlyDownloading = [];
+    var _filenamesOfImageToBeDownloaded = [];
+    var _exploration = null;
+    // imageLoadedCallback is an object of objects (identified by the filenames
+    // which are being downloaded at the time they are required by the
+    // directive).The object contains the resolve method of the promise
+    // attached with getInImageUrl method.
+    var _imageLoadedCallback = {};
+    var _recentlyRequestedImageFilenames = [];
```

Comment:
Why do we need to have this array? It seems to me that in all methods that work with this array the check that removeFromRecentlyRequestedImageFilenames contain filename is useless because we loaded the image just second ago or we know that the image is in cache.

Reply:
Actually it was being used in the previous implementation. I have removed it.

Done!